File: Guigna-RubyMotion/app/lib/systems/rudix.rb
# require 'GAdditions'
# require 'GLibrary'
# require 'GuignaItems'
# require 'GAgent'
# require 'GuignaSystems'

class Rudix < GSystem
  
  @prefix = '/usr/local'
  
  def initialize(agent=nil)
    super("Rudix", agent)
    @homepage = "http://rudix.org/"
    @cmd = "#{@prefix}/bin/rudix"
  end
  
  def clamped_os_version
    os_version = G.os_version
    if os_version < "10.6" || os_version > "10.9"
      os_version = "10.9"
    end
  end
  
  def list
    @index.clear
    @items.clear
    command = "#{cmd} search"
    osx_version = clamped_os_version()
    if G.os_version != osx_version
      command = "export OSX_VERSION=#{osx_version} ; #{cmd} search"
    end
    output = `#{command}`
    output.split("\n").each do |line|
      components = line.split "-"
      name = components[0]
      if components.size == 4
        name = name + "-" + components[1]
        components.delete_at(1)
      end
      version = components[1]
      version = version + "-" + components[2].split(".").first
      pkg = GPackage.new(name, version, self, :available)
      if !self[name].nil?
        prev_pkg = @items.delete self[name]
      end
      @items << pkg
      self[name] = pkg
    end
    self.installed # update index status
    @items
  end
    
  def installed
    if self.hidden?
      return @items.select {|pkg| pkg.status != :available}
    end
    pkgs = []
    return pkgs if self.mode == :online
    @items.each do |pkg|
      status = pkg.status
      pkg.installed = nil
      pkg.status = :available if status != :updated and status != :new
    end
    # self.outdated # update status of outdated packages
    output = `export HOME=~ ; export PATH=#{ENV["PATH"]} ; #{cmd}`
    output.split("\n").each do |line|
      name = line[line.rindex(".")+1..-1]
      pkg = self[name]
      latest_version = (pkg.nil? || pkg.version.nil?) ? nil : pkg.version.dup
      if pkg.nil?
        pkg = GPackage.new(name, latest_version, self, :uptodate)
        self[name] = pkg
      else
        if pkg.status == :available
          pkg.status = :uptodate
        end
      end
      pkg.installed = ""  # TODO
      pkgs << pkg
    end
    return pkgs
  end
      
  def refresh
    pkgs = []
    url = "http://rudix.org/download/2014/10.9/"
    links = agent.nodes_for_url(url, xpath:"//tbody//tr//a")
    decimalCharSet = NSCharacterSet.decimalDigitCharacterSet
    links.each do |link|
      name = link.stringValue
      next if name.start_with? 'Parent Dir' or name.include?("MANIFEST") or name.include?("ALIASES")
      idx = name.index "-"
      version = name[idx+1..-1]
      version = version[0...-4]
      if not decimalCharSet.characterIsMember(version[0].ord)
        idx2 = version.index("-")
        version = version[idx2+1..-1]
        idx += idx2+1
      end
      name = name[0...idx]
      pkg = GItem.new(name, version, self, :available)
      pkg.homepage = "http://rudix.org/packages/#{pkg.name}.html"
      pkgs << pkg
    end
    @items = pkgs
  end
  
  def home(item)
    "http://rudix.org/packages/#{item.name}.html"
  end
  
  def log(item)
    if !item.nil?
      "https://github.com/rudix-mac/rudix/commits/master/Ports/#{item.name}"
    else
      "https://github.com/rudix-mac/rudix/commits/"
    end
  end
  
  def contents(pkg)
    if !pkg.installed.nil?
      `export HOME=~ ; export PATH=#{ENV["PATH"]} ;  #{cmd} --files #{pkg.name}`
    else
      ""
    end
  end
  
  def cat(pkg)
    return NSString.stringWithContentsOfURL(NSURL.URLWithString("https://raw.githubusercontent.com/rudix-mac/rudix/master/Ports/#{pkg.name}/Makefile", encoding:NSUTF8StringEncoding, error:nil))
  end
  
  def install_cmd(pkg)
    command = "#{cmd} install #{pkg.name}"
    osx_version = clamped_os_version()
    if G.os_version != osx_version
      command = "OSX_VERSION=#{osx_version} #{command}"
    end
    "sudo #{command}"
  end
  def uninstall_cmd(pkg)
    "sudo #{cmd} remove #{pkg.name}"
  end
  
  def hide_cmd
    "sudo mv #{prefix} #{prefix}_off"
  end
  def unhide_cmd
    "sudo mv #{prefix}_off #{prefix}"
  end
  def self.setup_cmd
    "curl -s https://raw.githubusercontent.com/rudix-mac/rpm/master/rudix.py | sudo python - install rudix"
  end
  
end

File: Legacy/Guigna-Swift/Guigna/Systems/Rudix.swift
import Foundation

class Rudix: GSystem {

    override class var prefix: String { return "/usr/local" }

    init(agent: GAgent) {
        super.init(name: "Rudix", agent: agent)
        homepage = "http://rudix.org/"
        logpage = "https://github.com/rudix-mac/rudix/commits"
        cmd = "\(prefix)/bin/rudix"
    }

    class func clampedOSVersion() -> String {
        var osVersion = G.OSVersion()
        if osVersion < "10.6" || osVersion > "10.9" {
            osVersion = "10.9"
        }
        return osVersion
    }

    override func list() -> [GPackage] {

        index.removeAll(keepCapacity: true)
        items.removeAll(keepCapacity: true)

        var manifest = ""
        if mode == .Online { // FIXME: manifest is not available anymore
            manifest = (try? String(contentsOfURL: NSURL(string: "http://rudix.org/download/2014/10.9/00MANIFEST.txt")!, encoding: NSUTF8StringEncoding)) ?? ""
        } else {
            var command = "\(cmd) search"
            let osxVersion = Rudix.clampedOSVersion()
            if G.OSVersion() != osxVersion {
                command = "/bin/sh -c export__OSX_VERSION=\(osxVersion)__;__\(cmd)__search"
                manifest = output(command)
            }
        }
        var lines = manifest.split("\n")
        lines.removeLast()
        for line in lines {
            var components = line.split("-")
            var name = components[0]
            if components.count == 4 {
                name += "-\(components[1])"
                components.removeAtIndex(1)
            }
            var version = components[1]
            version += "-" + components[2].split(".")[0]
            let pkg = GPackage(name: name, version: version, system: self, status: .Available)
            if self[name] != nil {
                var found: Int?
                for (i, pkg) in items.enumerate() {
                    if pkg.name == name {
                        found = i
                        break
                    }
                }
                if let idx = found {
                    items.removeAtIndex(idx)
                }
            }
            items.append(pkg)
            self[name] = pkg
        }
        self.installed() // update status
        return items as! [GPackage]
    }


    override func installed() -> [GPackage] {

        if self.isHidden {
            return items.filter { $0.status != .Available} as! [GPackage]
        }

        var pkgs = [GPackage]()
        pkgs.reserveCapacity(50000)

        if mode == .Online {
            return pkgs
        }

        var outputLines = output("\(cmd)").split("\n")
        outputLines.removeLast()
        var status: GStatus
        for pkg in items as! [GPackage] {
            status = pkg.status
            pkg.installed = nil
            if status != .Updated && status != .New {
                pkg.status = .Available
            }
        }
        // self.outdated() // update status
        for line in outputLines {
            let name = line.substringFromIndex(line.rindex(".") + 1)
            var pkg: GPackage! = self[name]
            let latestVersion: String = (pkg == nil) ? "" : pkg.version
            if pkg == nil {
                pkg = GPackage(name: name, version: latestVersion, system: self, status: .UpToDate)
                self[name] = pkg
            } else {
                if pkg.status == .Available {
                    pkg.status = .UpToDate
                }
            }
            pkg.installed = "" // TODO
            pkgs.append(pkg)
        }
        return pkgs
    }


    override func home(item: GItem) -> String {
        for line in cat(item).split("\n") {
            if line.hasPrefix("Site=") {
                homepage = line.substringFromIndex(5).stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceAndNewlineCharacterSet())
                if homepage.hasPrefix("http") {
                    return homepage
                }
            }
        }
        return "http://rudix.org/packages/\(item.name).html"
    }


    override func log(item: GItem) -> String {
        return "https://github.com/rudix-mac/rudix/commits/master/Ports/\(item.name)"
    }

    override func contents(item: GItem) -> String {
        if item.installed != nil {
            return output("\(cmd) --files \(item.name)")
        } else {
            return "" // TODO: parse http://rudix.org/packages/%@.html
        }
    }

    override func cat(item: GItem) -> String {
        return (try? String(contentsOfURL: NSURL(string: "https://raw.githubusercontent.com/rudix-mac/rudix/master/Ports/\(item.name)/Makefile")!, encoding: NSUTF8StringEncoding)) ?? ""
    }


    override func installCmd(pkg: GPackage) -> String {
        var command = "\(cmd) install \(pkg.name)"
        let osxVersion = Rudix.clampedOSVersion()
        if G.OSVersion() != osxVersion {
            command = "OSX_VERSION=\(osxVersion) \(command)"
        }
        return "sudo \(command)"
    }

    override func uninstallCmd(pkg: GPackage) -> String {
        return "sudo \(cmd) remove \(pkg.name)"
    }

    override func fetchCmd(pkg: GPackage) -> String {
        var command = "cd ~/Downloads ; \(cmd) --download \(pkg.name)"
        let osxVersion = Rudix.clampedOSVersion()
        if G.OSVersion() != osxVersion {
            command = "cd ~/Downloads ; OSX_VERSION=\(osxVersion) \(cmd) --download \(pkg.name)"
        }
        return command
    }

    override var hideCmd: String! {
        get {
            return "sudo mv \(prefix) \(prefix)_off"
        }
    }

    override var unhideCmd: String! {
        get {
            return "sudo mv \(prefix)_off \(prefix)"
        }
    }

    class var setupCmd: String! {
        var command = "curl -s https://raw.githubusercontent.com/rudix-mac/rpm/master/rudix.py | sudo python - install rudix"
        let osxVersion = Rudix.clampedOSVersion()
        if G.OSVersion() != osxVersion {
            command = "curl -s https://raw.githubusercontent.com/rudix-mac/rpm/master/rudix.py | sudo OSX_VERSION=\(osxVersion) python - install rudix"
        }
        return command
    }

    class var removeCmd: String! {
        return "sudo /usr/local/bin/rudix -R" // TODO: prefix
    }

}

